Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -127,6 +125,24 @@ CFTypeID CGPathGetTypeID() {
 static Boolean __CGPathEqual(CFTypeRef cf1, CFTypeRef cf2) {
     __CGPath* path1 = (__CGPath*)cf1;
     __CGPath* path2 = (__CGPath*)cf2;
+
+    path1->closePath();
+    path2->closePath();
+
+    // ID2D1 Geometries have no isEquals method. However, for two geometries to be equal they are both reported to contain the other.
+    // Thus we must do two comparisons.
+
+    D2D1_GEOMETRY_RELATION relation = D2D1_GEOMETRY_RELATION_UNKNOWN;
+    if (SUCCEEDED(path1->_impl.pathGeometry->CompareWithGeometry(path2->_impl.pathGeometry.Get(), D2D1::IdentityMatrix(), &relation))) {
+        // Does path 1 contain path 2?
+        if (relation == D2D1_GEOMETRY_RELATION_IS_CONTAINED) {
+            if (SUCCEEDED(
+                    path2->_impl.pathGeometry->CompareWithGeometry(path1->_impl.pathGeometry.Get(), D2D1::IdentityMatrix(), &relation))) {
+                // Return true if path 2 also contains path 1
+                return (relation == D2D1_GEOMETRY_RELATION_IS_CONTAINED ? true : false);
+            }
+        }
+    }
```

Comment:
The nested logic in question is around the if(SUCCEEDED clauses. It is hard to read where the actual conditionals are admist those. A rewritten version is cleaner and looks like:

D2D1_GEOMETRY_RELATION relation = D2D1_GEOMETRY_RELATION_UNKNOWN;
RETURN_FALSE_IF_FAILED(path1->_impl.pathGeometry->CompareWithGeometry(path2->_impl.pathGeometry.Get(), D2D1::IdentityMatrix(), &relation)));

// Does path 1 contain path 2?
if (relation == D2D1_GEOMETRY_RELATION_IS_CONTAINED) {
    RETURN_FALSE_IF_FAILED(path2->_impl.pathGeometry->CompareWithGeometry(path1->_impl.pathGeometry.Get(), D2D1::IdentityMatrix(), &relation)));

    // Return true if path 2 also contains path 1
    return (relation == D2D1_GEOMETRY_RELATION_IS_CONTAINED ? true : false);
}


which makes the actual conditions around relation pop out.

---
In reply to: [85576878](https://github.com/Microsoft/WinObjC/pull/1255#discussion_r85576878) [](ancestors = 85576878,85566823)